### PR TITLE
Keep on top feature

### LINKED ISSUE #3038: Feature Request: Always On Top
# Description of the new feature/enhancement
Option to let terminal be always on top of all windows, just like the new calculator app. This would be useful for people learning how to do stuff, as it can be on top of the tutorial window while not suffering from text going on to the next line during resizing.

### PR DESCRIPTION
**Pull request by Chips1234**
## Summary of the Pull Request
Makes a button on the title bar to keep up on top of other windows. Calculator has a similar feature.
Please note, as of 06/13, the button design is finished but the code is not even started. This is a work in progress.  
 
## References
Connected to #3038.

## PR Checklist
* [X] Closes #3038
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [X] No documents

### CODE REVIEW

**Comment by Chips1234** [comment]
For some reason, there are always commits unrelated to this. So, please ignore the commits from Apr 17th, to May 24th. Only beginning on 30 June 2020.

**Comment by Chips1234** [comment]
Also, can someone help with the placement of the button? It looks good in the designer but it actually is placed on top of the "New Tab" split button

File: src/cascadia/TerminalApp/TitlebarControl.xaml [hunk 1]
```diff
@@ -36,6 +36,20 @@ the MIT License. See LICENSE in the project root for license information. -->
         Grid.Column="1"
         MinWidth="45.0"
         DoubleTapped="DragBar_DoubleTapped"/>
+    <Button
+        x:Name="KeepOnTopButton"
+        MinWidth="46.0"
+        Width="46.0"
+        Height="36"
+        Margin="220,0,0,0"
```
Comment (AnuthaDev) [inline src/cascadia/TerminalApp/TitlebarControl.xaml:44]: This should be removed. Generally you should not hardcode margins

File: src/cascadia/TerminalApp/TitlebarControl.xaml [hunk 2]
```diff
@@ -36,6 +36,20 @@ the MIT License. See LICENSE in the project root for license information. -->
...
+        Height="36"
+        Margin="220,0,0,0"
+        Content="&#xE840;"
+        FontFamily="Segoe MDL2 Assets"
+        FontSize="12"
+        FontWeight="SemiLight"
+        UseLayoutRounding="True"
+        HorizontalAlignment="Right"
+        VerticalAlignment="Stretch"
+        Grid.Row="2"/>
```
Comment (AnuthaDev) [inline src/cascadia/TerminalApp/TitlebarControl.xaml:52]: Replace it with Grid.Column="2"

**Comment by AnuthaDev** [comment]
> For some reason, there are always commits unrelated to this PR. So, please ignore the commits from Apr 17th, to May 24th. Only beginning on 30 June 2020.

Squash changes and force push :)

**Comment by AnuthaDev** [comment]
Also, this button should preferably be a part of the minmaxclosecontrol so it can reuse the styles and maybe even code

**Comment by AnuthaDev** [comment]
We might even want to make a separate control, so that (optional?) features like #5426 and #3038 can be implemented together

**Comment by Chips1234** [comment]
Alright, thanks!

**Comment by Chips1234** [comment]
> We might even want to make a separate control, so that (optional?) features like #5426 and #3038 can be implemented together

Alright.

**Comment by Chips1234** [comment]
If I set the value for the button for `Grid.column="2"` and for the `MinMaxClose Grid.Column="3"`, the button overlaps the minimize one.

**Comment by zadjii-msft** [comment]
Thanks for the effort! I think I'd prefer if we started by coming at this from another direction. I think it would be better to prove that we can get the Terminal to be set to "keep on top" mode first, then add the UI for enabling that setting. IMO, I'd rather this just be a setting that's enabled in the settings UI (#1564), or toggled with a keybinding/command, rather than have a dedicated button for it. So, if I were implementing this, I'd probably do it in the following steps:

1. First prove that I can get the Terminal to stay on top of all windows.
2. Then I'd add a setting to enable this fuctionality
3. THen I'd make sure that the setting can be hot-reloaded, so the value can be turned on/off at runtime. 
4. If a follow-up PR, I'd add a ShortcutAction to manually toggle that setting.

**Comment by AnuthaDev** [comment]
> IMO, I'd rather this just be a setting that's enabled in the settings UI (#1564), or toggled with a keybinding/command, rather than have a dedicated button for it.

I would prefer it like the calculator with a dedicated button in titlebar to change always on top status along with a keyboard shortcut to toggle the always on top flag. IMHO, you should rather add a setting to set whether or not to show the "always on top" button in titlebar

**Comment by AnuthaDev** [comment]
Also, it is interesting to think how this feature would interact with tab tearoff when it is implemented...

**Comment by zadjii-msft** [comment]
> Also, it is interesting to think how this feature would interact with tab tearoff when it is implemented...

That's exactly what [my question was](https://github.com/microsoft/terminal/issues/3038#issuecomment-537947136) when this issue was first posted:

> I suppose my next question would be: what happens when multiple applications want to be "always on top" - who wins?

and I'm not sure I've gotten a totally satisfactory answer to that question yet.

<hr>
> I would prefer it like the calculator with a dedicated button in titlebar to change always on top status along with a keyboard shortcut to toggle the always on top flag. IMHO, you should rather add a setting to set whether or not to show the "always on top" button in titlebar

I've been _highly_ reluctant to add more buttons to our already pretty minimalistic UI. It starts a slippery slope - "if this button is okay, then why can't I have a button dedicated to each profile"? I worry that adding more dedicated UI elements like this would leave the UI looking cluttered. Hence why I'm trying to drive the Command Palette (see #2046) for adding commands in a toggleable menu, and also why I'm driving to allow the user to add actions directly to the "New Tab Button dropdown" (see spec in-progress #5888). That seems like a much better place to stash this button, if the user so chooses. Hence my push to have this be a ShortcutAction the user can toggle 😉.

**Comment by WSLUser** [comment]
I just did some playing around using the Keep on Top feature in Calculator. I'll be honest, I'm not a huge fan of it. There's an ugly button right next to the Calculator Mode title and it always goes to the right. I actually prefer having it done exactly as zadjii-msft requests. Another thing to think about in regards to that "Shortcut Action": how will that play into Quake Mode? Will we be able to instantly place Terminal on top (toggable setting) when we launch it?

**Comment by AnuthaDev** [comment]
Also, should there be a maximum size limit in AOT mode?...

**Comment by ZhaoMJ** [comment]
Given PowerToys has a prototype of "always on top" tool, we might not need to implement it by ourselves. So we can keep our UI minimalistic. I also think they can figure out a solution for cases like multiple windows.

**Comment by WSLUser** [comment]
In some world where somebody chooses to use Windows Terminal but not Power Toys, we'd need this feature available as an option that defaults to false. I am a bit worried about conflicts between trying to use the same feature via both Terminal settings and PowerToys but I think that's simply a doc update that recommends users not to enable it in both but one or the other.

**Comment by zadjii-msft** [comment]
Thanks for the contribution! Since this PR seems to be a bit stale now, we're going to close this in favor of #6903. Thanks!

**Comment by Chips1234** [comment]
Alright, thanks

**Pull request closed without merging**